include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -250,118 +257,341 @@ __parallel_for(_ExecutionPolicy&& __exec, _Fp __brick, _Index __count, _Ranges&&
 }
 
 //------------------------------------------------------------------------
-// parallel_transform_reduce - async pattern
+// parallel_transform_reduce - async patterns
 //------------------------------------------------------------------------
-template <typename _Tp, ::std::size_t __grainsize = 4, typename _ExecutionPolicy, typename _Up, typename _LRp,
-          typename _Rp, typename _InitType,
-          oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy, int> = 0, typename... _Ranges>
+
+// Sequential parallel_transform_reduce used for small input sizes
+template <typename _Tp, typename _ReduceOp, typename _TransformOp, typename _ExecutionPolicy, typename _Size,
+          typename _InitType, oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy, int> = 0,
+          typename... _Ranges>
 auto
-__parallel_transform_reduce(_ExecutionPolicy&& __exec, _Up __u, _LRp __brick_leaf_reduce, _Rp __brick_reduce,
-                            _InitType __init, _Ranges&&... __rngs)
+__parallel_transform_reduce_seq_submitter(_ExecutionPolicy&& __exec, _Size __n, _ReduceOp __reduce_op,
+                                          _TransformOp __transform_op, _InitType __init, _Ranges&&... __rngs)
 {
-    auto __n = oneapi::dpl::__ranges::__get_first_range_size(__rngs...);
-    assert(__n > 0);
-
-    using _Size = decltype(__n);
     using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
     using _CustomName = typename _Policy::kernel_name;
     using _ReduceKernel =
-        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__reduce_kernel, _CustomName, _Up, _LRp,
-                                                                               _Rp, _Ranges...>;
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__reduce_seq_kernel, _CustomName,
+                                                                               _ReduceOp, _TransformOp, _Ranges...>;
+
+    auto __transform_pattern = unseq_backend::transform_init_seq<_ExecutionPolicy, _ReduceOp, _TransformOp>{
+        __reduce_op, _TransformOp{__transform_op}};
+    auto __reduce_pattern = unseq_backend::reduce<_ExecutionPolicy, _ReduceOp, _Tp>{__reduce_op};
+
+    sycl::buffer<_Tp> __res(sycl::range<1>(1));
+
+    sycl::event __reduce_event = __exec.queue().submit([&, __n](sycl::handler& __cgh) {
+        oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); // get an access to data under SYCL buffer
+        auto __res_acc = __res.template get_access<access_mode::write>(__cgh);
+        __cgh.single_task<_ReduceKernel>([=] {
+            _Tp __result;
```

Comment:
The line introduces an additional requirement for a type - a default  constructor.

Reply:
Thanks, I will change that.